jill/utils/version_utils.py

```diff
@@ -64,14 +65,25 @@ def get_full_version(version: str):
         return ".".join([major, minor, patch])
 
 
-def read_remote_json(url, cache=dict()):
+def read_remote_versions(upstream, cache=dict()):
     """
         If not cached, download and read json content from remote URL `url`.
     """
     if not cache:
+        registry = read_registry()
+        if upstream in registry:
+            versions_url = registry[upstream].versions_url
+        else:
+            # If not specified, use the first response as the result
+            # TODO: we can share the first_response result here so that later
+            #       downloading (e.g., `query_download_url`) does not need to check
+            #       it again.
+            versions_url_list = [x.versions_url for x in registry.values()]
+            versions_url = first_response(versions_url_list, timeout=0.5)
+            versions_url = versions_url if versions_url else DEFAULT_VERSIONS_URL
```

Comment:
If not specified `upstream` will be `None`, we'll then use the first URL that responded OK to download the `versions.json`. If we, unfortunately, can't even get a response until the timeout (default 3s), we use the hard-coded julialang versions.json as the fallback url.

To achieve this, I've added a new entry `versions` in "sources.json" for each upstream.